src/sentry/utils/snowflake.py

```diff
@@ -0,0 +1,116 @@
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import Tuple
+
+from django.conf import settings
+from django.db import IntegrityError
+
+from sentry.utils import redis
+
+_TTL = timedelta(minutes=5)
+SENTRY_EPOCH_START = datetime(2022, 4, 26, 0, 0).timestamp()
+
+
+class SnowflakeIdMixin:
+    def save_with_snowflake_id(self, snowflake_redis_key, save_callback):
+        for _ in range(settings.MAX_REDIS_SNOWFLAKE_RETRY_COUNTER):
+            if not self.id:
+                self.id = generate_snowflake_id(snowflake_redis_key)
+            try:
+                save_callback()
+                return
+            except IntegrityError:
+                self.id = None
+        raise Exception("Max allowed ID retry reached. Please try again in a second")
+
+
+@dataclass(frozen=True, eq=True)
+class SnowflakeBitSegment:
+    length: int
+    name: str
+
+    def __post_init__(self):
+        if self.length <= 0:
+            raise Exception("The length should be a positive number")
+
+    def validate(self, value):
+        if value >> self.length != 0:
+            raise Exception(f"{self.name} exceed max bit value of {self.length}")
+        return True
+
+
+BIT_SEGMENT_SCHEMA = (
+    VERSION_ID := SnowflakeBitSegment(5, "Version ID"),
+    TIME_DIFFERENCE := SnowflakeBitSegment(32, "Time difference"),
+    REGION_ID := SnowflakeBitSegment(12, "Region ID"),
+    REGION_SEQUENCE := SnowflakeBitSegment(4, "Region sequence"),
+)
+
+ID_VALIDATOR = SnowflakeBitSegment(
+    sum(segment.length for segment in BIT_SEGMENT_SCHEMA), "Snowflake ID"
+)
+assert ID_VALIDATOR.length == 53
+
+MAX_AVAILABLE_REGION_SEQUENCES = 1 << REGION_SEQUENCE.length
```

Comment:
@maxiuyuan can you add and assert that ` assert MAX_AVAILABLE_REGION_SEQUENCES > 0`?